roadmap/implementers-guide/src/node/disputes/dispute-distribution.md

```diff
@@ -159,114 +157,182 @@ session (which might be the same or not and may change during the dispute).
+
+## Reception
+
+As we shall see the receiving side is mostly about handling spam and ensuring
+the dispute-coordinator learns about disputes as fast as possible.
+
+Goals for the receiving side:
+
+1. Get new disputes to the dispute-coordinator as fast as possible, so
+  prioritization can happen properly.
+2. Batch votes per disputes as much as possible for good import performance.
+3. Prevent malicious nodes exhausting node resources by sending lots of messages.
+4. Prevent malicious nodes from sending so many messages/(fake) disputes,
+  preventing us from concluding good ones.
+5. Limit ability of malicious nodes of delaying the vote import due to batching
+   logic.
+
+Goal 1 and 2 seem to be conflicting, but an easy compromise is possible: When
+learning about a new dispute, we will import the vote immediately, making the
+dispute coordinator aware and also getting immediate feedback on the validity.
+Then if valid we can batch further incoming votes, with less time constraints as
+the dispute-coordinator already knows about the dispute.
+
+Goal 3 and 4 are obviously very related and both can easily be solved via rate
+limiting as we shall see below. Rate limits should already be implemented at the
+substrate level, but [are not](https://github.com/paritytech/substrate/issues/7750)
+at the time of writing. But even if they were, the enforced substrate limits would
+likely not be configurable and thus would still be to high for our needs as we can
+rely on the following observations:
+
+1. Each honest validator will only send one message (apart from duplicates on
+  timeout) per candidate/dispute.
+2. An honest validator needs to fully recover availability and validate the
+  candidate for casting a vote.
+
+With these two observations, we can conclude that honest validators will usually
+not send messages at a high rate. We can therefore enforce conservative rate
+limits and thus minimize harm spamming malicious nodes can have.
+
+Before we dive into how rate limiting solves all spam issues elegantly, let's
+further discuss that honest behaviour further:
+
+What about session changes? Here we might have to inform a new validator set of
+lots of already existing disputes at once.
+
+With observation 1) and a rate limit that is per peer, we are still good:
+
+Let's assume a rate limit of one message per 200ms per sender. This means 5
+messages from each validator per second. 5 messages means 5 disputes!
+Conclusively, we will be able to conclude 5 disputes per second - no matter what
+malicious actors are doing. This is assuming dispute messages are sent ordered,
+but even if not perfectly ordered: On average it will be 5 disputes per second.
+
+This is good enough! All those disputes are valid ones and will result in
+slashing. Let's assume all of them conclude `valid`, and we slash 1% on those.
+This will still mean that nodes get slashed 100% in just 20 seconds.
```

Comment:
As mentioned in the chat, this is not how slashes are applied: https://github.com/paritytech/substrate/blob/22b678f7284537b28c0983d6301faea3c607278c/frame/staking/src/slashing.rs#L255-L261. It would still be 1%. However, we can implement exponential slashing or https://github.com/paritytech/polkadot/issues/5898#issuecomment-1221188631. I'm hesitant to implement slashing for valid disputes up to 100% though, because it practice, if a validator is slow it will dispute several candidates, not just one. I think something like 5-10% as an upper bound plus disabling makes more sense in case of repeated disputes against valid candidates.

Reply:
By disabling I mean ignoring dispute votes for X sessions (both off-chain and on-chain), kicking out of validator set on the next session (modulo https://github.com/paritytech/substrate/issues/5543) and ignoring blocks produced by them.

Reply:
Yes from the perspective of dispute-distribution and dispute-coordinator: Disabling is more important than slashing. I'll adjust the guide accordingly.